pipeline/data_store.py

```diff
@@ -0,0 +1,773 @@
+            raise ValueError(f'image must be an Image object, got {type(value)}')
+
+        if key == 'sources' and not isinstance(value, SourceList):
+            raise ValueError(f'sources must be a SourceList object, got {type(value)}')
+
+        if key == 'wcs' and not isinstance(value, WorldCoordinates):
+            raise ValueError(f'WCS must be a WorldCoordinates object, got {type(value)}')
+
+        if key == 'zp' and not isinstance(value, ZeroPoint):
+            raise ValueError(f'ZP must be a ZeroPoint object, got {type(value)}')
+
+        if key == 'ref_image' and not isinstance(value, Image):
+            raise ValueError(f'ref_image must be an Image object, got {type(value)}')
+
+        if key == 'sub_image' and not isinstance(value, Image):
+            raise ValueError(f'sub_image must be a Image object, got {type(value)}')
+
+        if key == 'detections' and not isinstance(value, SourceList):
+            raise ValueError(f'detections must be a SourceList object, got {type(value)}')
+
+        if key == 'cutouts' and not isinstance(value, list):
+            raise ValueError(f'cutouts must be a list of Cutout objects, got {type(value)}')
+
+        if key == 'cutouts' and not all([isinstance(c, Cutouts) for c in value]):
+            raise ValueError(f'cutouts must be a list of Cutouts objects, got list with {[type(c) for c in value]}')
+
+        if key == 'measurements' and not isinstance(value, list):
+            raise ValueError(f'measurements must be a list of Measurements objects, got {type(value)}')
+
+        if key == 'measurements' and not all([isinstance(m, Measurements) for m in value]):
+            raise ValueError(
+                f'measurements must be a list of Measurement objects, got list with {[type(m) for m in value]}'
+            )
+
+        if key == 'upstream_provs' and not isinstance(value, list):
+            raise ValueError(f'upstream_provs must be a list of Provenance objects, got {type(value)}')
+
+        if key == 'upstream_provs' and not all([isinstance(p, Provenance) for p in value]):
+            raise ValueError(
+                f'upstream_provs must be a list of Provenance objects, got list with {[type(p) for p in value]}'
+            )
+
+        if key == 'session' and not isinstance(value, (sa.orm.session.Session, SmartSession)):
+            raise ValueError(f'Session must be a SQLAlchemy session or SmartSession, got {type(value)}')
+
+        super().__setattr__(key, value)
+
+    def get_inputs(self):
+        """Get a string with the relevant inputs. """
+
+        if self.im_id is not None:
+            return f'im_id={self.im_id}'
+        elif self.exp_id is not None and self.ccd_id is not None:
+            return f'exp_id={self.exp_id}, ccd_id={self.ccd_id}'
+        else:
+            raise ValueError('Could not get inputs for DataStore.')
+
+    def get_provenance(self, process, pars_dict, upstream_provs=None, session=None):
```

Comment:
So, yeah, this is the way I'd do it.  Pass a session to each method, which it uses as the argument to SmartSesssion.  Don't save it in the DataStore self.

Reply:
As I say above, this is exactly what I wanted, but the arguments are passed in through the data store, there's no way to avoid putting that session in the ds.

Reply:
I don't understand why the session has to be in the datastore.  Indeed, as I said above, I think a design that requires that goes against the plan to only open sessions when they're needed.